src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -455,6 +455,60 @@ contract NiftyswapExchange20 is ReentrancyGuard, ERC1155MintBurn, INiftyswapExch
     emit LiquidityAdded(_provider, _tokenIds, _tokenAmounts, currencyAmounts);
   }
 
+  /**
+   * @dev Convert pool participation into amounts of token and currency.
+   * @dev Rounding error of the asset with lower resolution is traded for the other asset.
+   * @param _amountPool       Participation to be converted to tokens and currency.
+   * @param _tokenReserve     Amount of tokens on the AMM reserve.
+   * @param _currencyReserve  Amount of currency on the AMM reserve.
+   * @param _totalLiquidity   Total liquidity on the pool.
+   *
+   * @return currencyAmount Currency corresponding to pool amount plus rounded tokens.
+   * @return tokenAmount    Token corresponding to pool amount plus rounded currency.
+   */
+    function _toRoundedLiquidity(
+    uint256 _amountPool,
+    uint256 _tokenReserve,
+    uint256 _currencyReserve,
+    uint256 _totalLiquidity
+  ) internal pure returns (
+    uint256 currencyAmount,
+    uint256 tokenAmount,
+    int256  tradedCurrency,
+    int256  tradedToken
+  ) {
+    uint256 currencyProduct = _amountPool.mul(_currencyReserve);
+    uint256 tokenProduct = _amountPool.mul(_tokenReserve);
+
+    if (tokenProduct > currencyProduct) {
```

Comment:
I'm inclining more and more on leaving this up to the client. I feel like it's hard to predict what users will want and leaving it up to the client would give us more freedom. Current situation could lead to large slippage, say if you round to 0 from 0.99 instead of rounding to 1. Fortunately leaving it up to the client doesn't change too much the code you have here, just an extra argument. 

Could be a boolean called `_roundingSellToken` where `false` means you will buy the token to avoid rounding errors and `true` would mean you would sell the rounding error tokens into currency. I think buying the token will be more common than selling it, hence why false is the default (a bit less calldata).

Reply:
Fair enough, I didn't want to change the interface, but it could give us more flexibility for the future.

Reply:
Is not the exact same code though, because the rounding error on currency is probably not enough to "complete" the missing amount for buying a full token, so the code needs to use "full" currency units beyond just the rounding error.

This is fine sometimes, but in some cases the currency available may not be enough to complete the token... then the whole process should fail, given that the user doesn't have enough currency for the trade (so we need to allow for the alternative).

It's possible to implement this, but I worry it would increase the complexity of the process quite a bit. For example, it requires to determine how many currency/tokens are going to be converted to the other side, (just the rounding error?, enough to negate the other rounding error?, an x amount?, etc), this bloats the method a bit too much, and the contract is already close to the 24KB contract limit.

If I had to pick a way I would make the amounts to be traded explicit as an input, this is not ideal either, because if there are trades in between the transaction construction and the confirmation it could break those pre-defined parameters, maybe we may even need to build a router contract on top of this.

Another issue is that letting the user configure the behavior increases the surface of attack, in the sense that the user has more inputs to manipulate when calling the contract, and thus a higher chance of finding an vulnerability on the mechanism.

In my opinion in this case is better to keep it as simple as possible, the good thing about the current implementation is that it only trades the rounding error, thus the user is not exposed to loosing funds during withdrawal, and surface of attack should also be smaller.